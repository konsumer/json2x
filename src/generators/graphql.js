import capitalize from 'capitalize'
import { getTypes } from '../index'

export const description = 'Generate GraphQL types from your data'

export const options = yargs => {
  yargs.option('input', {description: 'Generate input-types too', alias: 'i'})
  yargs.example('$0 graphql User ./users/*.json', 'Make a type called User from a bunch of JSON files.')
}

export const makeType = (name, types, input) => {
  const others = []
  name = input && name.substr(0, 5) !== 'Input' ? `Input${name}` : name
  return `${input ? `input ${name}` : `type ${name}`} {
  ${Object.keys(types).map(t => {
    if (typeof types[t] === 'string') {
      return `${t}: ${capitalize(types[t])}`
    } else {
      if (Array.isArray(types[t])) {
        return `${t}: [${capitalize(types[t][0])}]`
      } else {
        others.push(makeType(`${capitalize(name)}${capitalize(t)}`, types[t], input))
        return `${t}: ${capitalize(name)}${capitalize(t)}`
      }
    }
  }).join('\n  ')}
}
` + others.join('')
}

export const run = ({ MODEL, data, input }) => {
  const types = getTypes(Object.values(data))
  let out = makeType(MODEL, types)
  if (input) {
    out += makeType(MODEL, types, true)
  }
  return out
}
