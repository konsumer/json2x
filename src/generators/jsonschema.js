import jsonSchemaGenerator from 'json-schema-generator'

// TODO: use smarter parser to derive schema from multiple records

export const description = 'Generate JSONSchema from your data'

export const run = ({ MODEL, ast, data }) => {
  const schema = jsonSchemaGenerator(Object.values(data)[0])
  schema.description = MODEL
  return JSON.stringify(schema, null, 2)
}
