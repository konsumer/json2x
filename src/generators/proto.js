export const description = 'Generate protobuf message-definitions from your data'

export const run = ({MODEL}, ast) => {
  console.log({MODEL})
}
