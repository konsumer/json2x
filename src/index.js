export const allArraySame = obj => {
  for (var i = 1; i < obj.length; i++) {
    if (obj[i] !== obj[0]) return false
  }
  return true
}

export const isFloat = n => n === +n && n !== (n | 0)

export const getBestType = obj => {
  const t = ({}).toString.call(obj).split(' ')[1].slice(0, -1).toLowerCase()
  switch (t) {
    case 'object': return Array.isArray(obj) ? 'array' : 'object'
    case 'number': return isFloat(obj) ? 'float' : 'int'
    default: return t
  }
}

const getAverageFromTypes = types => {
  if (allArraySame(types)) {
    return types[0]
  }
  // if any are a string, return string
  if (types.indexOf('string') !== -1) {
    return 'string'
  }
  // if any are a float, return float
  if (types.indexOf('float') !== -1) {
    return 'float'
  }
  return types[0]
}

export const getAverageType = values => getAverageFromTypes(values.map(getBestType))

/**
 * Get types from an Object filled with similar objects
 * @param  object data An Object filled with similar objects
 * @return object      Type info
 */
export const getTypes = data => {
  const fields = {}
  const out = {}
  const objs = {}
  const arrays = {}
  data.forEach(record => {
    Object.keys(record).forEach(k => {
      const type = getBestType(record[k])
      if (type === 'array') {
        arrays[k] = arrays[k] || []
        return arrays[k].push(getAverageType(record[k]))
      }
      if (type === 'object') {
        objs[k] = objs[k] || []
        return objs[k].push(record[k])
      }
      fields[k] = fields[k] || []
      return fields[k].push(record[k])
    })
  })

  Object.keys(fields).forEach(k => {
    out[k] = getAverageType(fields[k])
  })

  Object.keys(objs).forEach(k => {
    out[k] = getTypes(objs[k])
  })

  Object.keys(arrays).forEach(k => {
    out[k] = [getAverageFromTypes(arrays[k])]
  })

  return out
}

module.exports = { allArraySame, getBestType, getAverageType, getTypes }
