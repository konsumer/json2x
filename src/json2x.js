#!/usr/bin/env babel-node

import { promisify } from 'util'
import { basename } from 'path'
import { readFileSync } from 'fs'
import yargs from 'yargs'
import globR from 'glob-require'

require('babel-polyfill')

const noop = () => {}
const g = promisify(globR)

const main = async () => {
  const commands = await g('**/*.js', {cwd: `${__dirname}/generators`})

  commands.forEach(({path, exports: {description, options, run}}) => {
    const name = basename(path, '.js')
    yargs.command(
      `${name} <MODEL> <FILES...>`,
      description,
      options || noop,
      opt => {
        opt.raw = {}
        opt.data = {}
        opt.FILES.forEach(source => {
          opt.raw[source] = readFileSync(source).toString()
          opt.data[source] = JSON.parse(opt.raw[source])
        })
        Promise.resolve(run(opt)).then(results => console.log(results))
      }
    )
    yargs.example(`$0 ${name} -h`, `Get some help with the ${name} command`)
  })

  yargs // eslint-disable-line
    .demandCommand()
    .help()
    .version()
    .alias('h', 'help')
    .alias('v', 'version')
    .wrap(process.stdout.columns)
    .argv
}
main()
