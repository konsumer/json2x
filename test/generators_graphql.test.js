/* global describe, it, expect */
import * as gen from '../src/generators/graphql'
import { g } from '../src/json2x'

describe('generators/graphql', () => {
  it('should have the expected description', () => {
    expect(gen.description).toMatchSnapshot()
  })

  it('should output the expected graphql schema for the mock', async () => {
    const data = (await g('*.json', {cwd: `${__dirname}/mock`})).map(({ exports }) => exports)
    const underTest = await gen.run({ MODEL: 'TestType', data, input: false })
    expect(underTest).toMatchSnapshot()
  })

  it('should output the expected graphql input schema for the mock', async () => {
    const data = (await g('*.json', {cwd: `${__dirname}/mock`})).map(({ exports }) => exports)
    const underTest = await gen.run({ MODEL: 'TestInput', data, input: true })
    expect(underTest).toMatchSnapshot()
  })
})
