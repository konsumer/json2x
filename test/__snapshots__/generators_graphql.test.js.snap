// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generators/graphql should have the expected description 1`] = `"Generate GraphQL types from your data"`;

exports[`generators/graphql should output the expected graphql input schema for the mock 1`] = `
"type TestInput {
  name: String
  username: String
  email: String
  phone: String
  website: String
  id: String
  lastSaid: String
  doing: String
  address: TestinputAddress
  company: TestinputCompany
  testnumobjnums: TestinputTestnumobjnums
}
type TestinputAddress {
  street: String
  suite: String
  city: String
  zipcode: String
  geo: TestinputaddressGeo
}
type TestinputaddressGeo {
  lat: Float
  lng: Float
}
type TestinputCompany {
  name: String
  catchPhrase: String
  bs: String
}
type TestinputTestnumobjnums {
  num1: Int
  num2: Int
  num3: Float
  deeper: TestinputtestnumobjnumsDeeper
}
type TestinputtestnumobjnumsDeeper {
  arr: [Float]
}
input InputTestInput {
  name: String
  username: String
  email: String
  phone: String
  website: String
  id: String
  lastSaid: String
  doing: String
  address: InputtestinputAddress
  company: InputtestinputCompany
  testnumobjnums: InputtestinputTestnumobjnums
}
input InputtestinputAddress {
  street: String
  suite: String
  city: String
  zipcode: String
  geo: InputtestinputaddressGeo
}
input InputtestinputaddressGeo {
  lat: Float
  lng: Float
}
input InputtestinputCompany {
  name: String
  catchPhrase: String
  bs: String
}
input InputtestinputTestnumobjnums {
  num1: Int
  num2: Int
  num3: Float
  deeper: InputtestinputtestnumobjnumsDeeper
}
input InputtestinputtestnumobjnumsDeeper {
  arr: [Float]
}
"
`;

exports[`generators/graphql should output the expected graphql schema for the mock 1`] = `
"type TestType {
  name: String
  username: String
  email: String
  phone: String
  website: String
  id: String
  lastSaid: String
  doing: String
  address: TesttypeAddress
  company: TesttypeCompany
  testnumobjnums: TesttypeTestnumobjnums
}
type TesttypeAddress {
  street: String
  suite: String
  city: String
  zipcode: String
  geo: TesttypeaddressGeo
}
type TesttypeaddressGeo {
  lat: Float
  lng: Float
}
type TesttypeCompany {
  name: String
  catchPhrase: String
  bs: String
}
type TesttypeTestnumobjnums {
  num1: Int
  num2: Int
  num3: Float
  deeper: TesttypetestnumobjnumsDeeper
}
type TesttypetestnumobjnumsDeeper {
  arr: [Float]
}
"
`;
